software/contrib/harmonic_lfos.py

```diff
@@ -41,8 +42,27 @@ def __init__(self):
 
         # Set the digital input and button handlers
         din.handler(self.reset)
-        b1.handler(self.change_mode)
-        b2.handler(self.increment_selection)
+
+        @b1.handler_falling
+        def b1Pressed():
+            """Triggered when B1 is pressed"""
+            if (
+                ticks_diff(ticks_ms(), b1.last_pressed()) > 1000
+                and ticks_diff(ticks_ms(), b1.last_pressed()) < 5000
+            ):
+                """Long press: toggle waveform view mode"""
+                self.viewAllWaveforms = not self.viewAllWaveforms
+            elif ticks_diff(ticks_ms(), b1.last_pressed()) < 500:
+                """Short press: Change the mode that controls wave shape"""
+                self.modes[self.selected_lfo] = (self.modes[self.selected_lfo] + 1) % self.MODES_COUNT
```

Comment:
`[Recommended][Question]`

A few suggestions/questions:

1) Do you need to check for the maximum duration of the long-press? You could simplify the logic by just treating anything > 1000ms as a long-press

2) If you do want to limit the long-press to 5000ms, instead of calling `ticks_ms` and `ticks_diff` multiple times, it would be more efficient to call them once and assign the difference to a local variable.

Generally function calls inside an interrupt handler are best avoided as much as possible; they increase the execution time and increase the possibility that you miss another interrupt.

I'd recommend refactoring this handler to be something like this:

```python
diff = ticks_diff(ticks_ms(), b1.last_pressed())
if diff > LONG_PRESS_MIN_DURATION and diff < LONG_PRESS_MAX_DURATION:
    # long-press
elif diff < SHORT_PRESS_MAX_DURATION:
    # short-press
```

If you don't need to limit the maximum duration, you can replace the end with a simple `else` and get rid of the `and` in the first section, further simplifying things.

Reply:
i like the way you suggest checking the duration better. I will implement that.

I also like having the max duration. maybe it’s just me, but things get bumped in my case. Sometimes your button sticks because the panel you made doesn’t fit perfectly 😑 or maybe you didn’t notice something resting on the button. it would just be nice to not have the screen toggle views once you realize this and unpress the button.

Maybe it’s just me and the LONG_PRESS_MAX_DURATION check is unnecessary.

Reply:
I think the maximum duration for a long press is subjective. I don't have any problems keeping it in, but it might be helpful in the documentation to specify the durations of a long & short press.  I think it's unlikely anyone will hold the button for more than 5s for a long-press, but just in case.

Since all the long-press does is change the visualization, and doesn't actually change the frequencies of anything, I don't think accidentally triggering it because of a cable pressing on the button is going to be the end of the world if it does happen.  As for buttons getting stuck in panels, that's more of an assembly defect, and not necessarily something we need to program around IMO.  The assumption that the button is usable and in a good state of repair doesn't feel unreasonable to me.